File: Graph_1203/jiqing/200.java
class Solution {
    int m, n;
    int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
    public int numIslands(char[][] grid) {
        m = grid.length;
        n = grid[0].length;
        int sum = 0;
        
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(grid[i][j] == '1') {
                    dfs(grid, i, j);
                    sum++;
                }
            }
        }
        
        return sum;

    }
    
    private void dfs(char[][] grid, int x, int y) {
        if(x < 0 || x >= m || y < 0 || y >= n || grid[x][y] == '0') return;
        
        grid[x][y] = '0';
        
        for(int i = 0; i < 4; i++) {
            int newx = x + directions[i][0];
            int newy = y + directions[i][1];
            
            dfs(grid, newx, newy);
        }
        return;
    }
}

File: Graph_1203/zzx/200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        directions=[[0,1],[0,-1],[1,0],[-1,0]]
        m=len(grid)
        n=len(grid[0])
        def dfs(i,j):
            grid[i][j]="0"
            for xx,yy in directions:
                if 0<=i+xx<m and 0<=j+yy<n and grid[i+xx][j+yy]=="1":
                    dfs(i+xx,j+yy)
        res=0
        for i in range(m):
            for j in range(n):
                if grid[i][j]=="1":
                    dfs(i,j)
                    res+=1
        return res
                    
            
        
        